backend/server/db/helpers/models.py

```diff
@@ -0,0 +1,115 @@
+from typing import Dict, List, Literal, NewType, Optional, Union
+from uuid import UUID
+from pydantic import BaseModel
+
+#
+# NewTypes
+#
+SessionID = NewType('SessionID', UUID)
+SessionToken = NewType('SessionToken', str)
+RefreshToken = NewType('RefreshToken', str)
+
+#
+# User Storage Models
+#
+class UserDegreeStorage(BaseModel):
+    programCode: str
+    specs: List[str]
+
+
+class UserCourseStorage(BaseModel):
+    code: str
+    suppressed: bool
+    mark: Union[Literal['SY', 'FL', 'PS', 'CR', 'DN', 'HD'], int, None]
+    uoc: int
+    ignoreFromProgression: bool
+
+# TODO-OLLI(pm): https://docs.pydantic.dev/latest/concepts/models/#rootmodel-and-custom-root-types
+# we can make this a RootModel or use a TypeAdapter if we want stronger type checks here
+type UserCoursesStorage = Dict[str, UserCourseStorage]
+
+class YearTerm(BaseModel):
+    Y: int
+    T: int
+
+class PlannerYear(BaseModel):
+    T0: List[str]
+    T1: List[str]
+    T2: List[str]
+    T3: List[str]
+
+class UserPlannerStorage(BaseModel):
+    unplanned: List[str]
+    startYear: int
+    isSummerEnabled: bool
+    mostRecentPastTerm: YearTerm
+    years: List[PlannerYear]
+    lockedTerms: Dict[str, bool]
+
+class _BaseUserStorage(BaseModel):
+    # NOTE: could also put uid here if we want
+    guest: bool
+
+class UserStorage(_BaseUserStorage):
+    setup: Literal[True] = True
+    degree: UserDegreeStorage
+    courses: UserCoursesStorage
+    planner: UserPlannerStorage
+
+class NotSetupUserStorage(_BaseUserStorage):
+    setup: Literal[False] = False
+
+# https://github.com/pydantic/pydantic/issues/1223#issuecomment-1152323275
+class PartialUserStorage(BaseModel):
+    # for batch user storage update
+    # TODO-OLLI(pm): remove this and use keyword args since they cannot be None
```

Comment:
There is a way for uvicorn to handle this 👀
We were forced to do this because BaseModel didn't allow "NotRequired"

Reply:
how do u mean?